keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -0,0 +1,619 @@
+* Allow cluster managers to provide a method to obtain credentials that doesn't require to be embedded into the controller code
+  and recompiling.
+* Be a secure mechanism for credential obtention and storage.
+
+### Non-Goals
+
+* Define the mechanism for shipping plugins to be used by the controllers
+* Design plugin or a library for plugins
+* Mandate Federated workload identity / OIDC frameworks (though they are recommended)
+
+## Proposal
+
+The proposed approach to issuing credentials is to leverage plugins for credentials, where a controller could leverage a library running local code
+which would retrieve the credentials for the current controller and for the clusterprofile of their choice. It is expected that plugins would
+leverage the identity of the controller (for example, the Kubernetes Service Account when running in Kubernetes) to retrieve credentials that are valid on other clusters.
+Plugins would be exec'ed by the controller so that they don't need be built-in the binary, allowing cluster managers to
+write their own credentials and still leveraging multicluster controllers written by the community.
+In addition, we propose to reuse the exec approach and protocol used for external credentials in Kubeconfig. Finally,
+in order to retrieve the endpoint for the cluster, we standardize the property names that are used in ClusterProfile.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The proposal's implementation would be done via a Library in https://github.com/kubernetes-sigs/cluster-inventory-api.
+The library would be in golang. The library is provided as the community shared implementation for golang and
+it is possible that other implementations would be created, and would work with the same plugin mechanism defined here,
+allowing for reuse of the external providers that cluster managers write.
+
+The expected prototype for a controller is expected to be the following:
+
+`func (c *ClusterProfileExternalProviders) GetConfig(cp ClusterProfile) (rest.Config, error)`
+
+The library implementation flow is expected to be as follow:
+
+1. Build the endpoint details of the cluster by reading properties of the ClusterProfile
+2. Call the CredentialsExternalProviders, following the same flow defined in [KEP 541](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md)
+3. Build the rest.Config and return it to the caller
+
+### Converting endpoint properties to cluster object
+
+The Cluster structure for the exec defined in KEP 541 assumes the following:
```

Comment:
I think we need to have an additional free form "data" in the clusterProfile API so that each "plugin" can utilize during the authentication process. i.e. Azure plugin would need the cluster's tenant and subscription ID.